usersession/userd/launcher.go

```diff
@@ -123,6 +130,126 @@ func (s *Launcher) OpenURL(addr string, sender dbus.Sender) *dbus.Error {
 	return nil
 }
 
+// OpenDesktopEntryEnv implements the 'OpenDesktopEntryEnv' method of the 'io.snapcraft.Launcher'
+// DBus interface.
+func (s *Launcher) OpenDesktopEntryEnv(desktop_file_id string, env []string, sender dbus.Sender) *dbus.Error {
+	desktop_file, err := desktopFileIdToFilename(desktop_file_id)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	exec_command, err := readExecCommandFromDesktopFile(desktop_file)
+
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	args, err := shlex.Split(exec_command)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	// Passing exec variables between confined snaps raises unanswered questions and they are not required
+	// for the simple cases.  For now, we don't have support for passing them in the dbus API and drop
+	// them from the command.
+	// https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#exec-variables
+	i := 0
+	for {
+		if strings.HasPrefix(args[i], "%") {
+			args = append(args[:i], args[i+1:]...)
+		} else {
+			i++
+		}
+
+		if i == len(args) {
+			break
+		}
+	}
+
+	cmd := exec.Command(args[0], args[1:]...)
+	cmd.Env = os.Environ()
+	for _, e := range env {
+		cmd.Env = append(cmd.Env, e)
+	}
+
+	if err := cmd.Start(); err != nil {
+		return dbus.MakeFailedError(fmt.Errorf("cannot run %q", exec_command))
+	}
+
+	return nil
+}
+
+// findDesktopFile recursively tries each subdirectory that can be formed from the (split) desktop file ID.
+func findDesktopFile(base_dir string, splitFileId []string) (string, error) {
+	desktop_file := filepath.Join(base_dir, strings.Join(splitFileId, "-"))
+	fileStat, _ := os.Stat(desktop_file)
+
+	if fileStat != nil {
+		return desktop_file, nil
+	}
+
+	for i := 1; i != len(splitFileId)-1; i++ {
+		desktop_file, err := findDesktopFile(filepath.Join(base_dir, strings.Join(splitFileId[:i], "-")), splitFileId[i:])
+		if err == nil {
+			return desktop_file, nil
+		}
+	}
+
+	return "", fmt.Errorf("cannot find desktop file")
+}
+
+// desktopFileIdToFilename determines the path associated with a desktop file ID.
+func desktopFileIdToFilename(desktop_file_id string) (string, error) {
+
+	// Currently the caller only has access to /var/lib/snapd/desktop/applications/, so we just look there
+	// and ignore https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html
+	base_dir := "/var/lib/snapd/desktop/applications/"
+
+	desktop_file, err := findDesktopFile(base_dir, strings.Split(desktop_file_id, "-"))
+
+	if err == nil {
```

Comment:
The error path is unusual. If there's an error it is customary to return the error right there. If there is no error we proceed onward to return the desktop file. In addition the actual error from `findDesktopFile` is lost, as the error path returns a new instance without relationship to the true cause.

Reply:
Ack, I shouldn't be using `error` to fudge a fallible/optional return value.